accel/tcg/cputlb.c

```diff
@@ -1214,11 +1226,26 @@ void tlb_set_page_with_attrs(CPUState *cpu, target_ulong vaddr,
      */
     desc->iotlb[index].addr = iotlb - vaddr_page;
 #ifdef TARGET_CHERI
-    // Cache the CHERI tag block. This massively speeds up running QEMU: before
-    // we added this optimization 10-25% of total runtime could be spent
-    // looking up tag blocks for a given virtual address.
-    desc->iotlb[index].tagmem =
-        cheri_tagmem_for_addr(section->mr->ram_block, xlat, size);
+    /*
+     * Cache the CHERI tag block. This massively speeds up running QEMU: before
+     * we added this optimization 10-25% of total runtime could be spent
+     * looking up tag blocks for a given virtual address.
+     */
+    desc->iotlb[index].tagmem_write = desc->iotlb[index].tagmem_read = tagmem;
```

Comment:
Pardon a naive question, but: if these are always set to the same value and differ only by the flags below, would it be better to store them as `.tagmem` and `.tagmem_flags`?  This would let `get_tagmem_from_iotlb_entry` not need its `bool write` parameter and would leave flag checking to the caller and doesn't close the door on `NULL` vs. `ALL_ZERO_TAGBLK` special cases, too, I think?

There's an assertion of simplified logic, below, but I seem to be missing something.  Perhaps expand that comment?

Reply:
Eventually I plan to move these fields to the non-IO TLB, where read/write entries are kept separate. I think this is to simplify access from TCG, so one entry can allow read/write, and the other can trap, without needing to be disentangled. Possibly this is not needed for the tagmem as well.